File: packages/sdk/src/graph.ts
// ─── Graph — declarative multi-step workflows ──────────────────────────────
//
// Lightweight orchestrator that sits on top of `Agent` + `subagent` + the
// existing run-linking infrastructure. Lives entirely in the SDK; the
// server still sees the resulting runs as a tree linked by workflow_id.
//
// Why bother with a Graph API at all when `subagent()` already composes
// agents? Two reasons:
//
//   1. State passing. With subagent, the parent LLM decides when and how
//      to call the child via tool_use. With a graph, the developer wires
//      the data flow explicitly — useful when the steps are deterministic
//      and you want predictable cost.
//
//   2. Visualization. A declarative graph is something the dashboard can
//      render before any run happens. A `subagent` tree only exists
//      after the LLM has decided to call it.
//
// Design choices kept deliberately small:
//
//   - No DAG validation at definition time beyond "edge target must exist
//     and START/END are special". Cycles are allowed; we cap with
//     maxSteps to prevent runaway.
//   - State is a plain object; steps return a partial that's merged in.
//   - Each step run becomes its own Relay run (so it shows up in traces
//     and counts toward the workflow cost).
//   - Errors bubble. No retry policy here — if you want retries, wrap
//     your step.

import { createAgent, type Agent } from "./agent.js";
import type { AgentEvent } from "./types.js";

export const START = "__start__" as const;
export const END = "__end__" as const;

type NodeName = string;
type EdgeTarget = NodeName | typeof END;

export type StepContext = {
  // Run id of the workflow root, present for every step in the workflow.
  workflowId: string;
  // The step name currently executing.
  stepName: string;
  // Mutate-safe view of the accumulated state. Returning a partial
  // merges; mutating in place is allowed but discouraged.
  signal?: AbortSignal;
};

export type StepFn<S extends Record<string, unknown>> = (
  state: S,
  ctx: StepContext,
) => Promise<Partial<S>> | Partial<S>;

export type ConditionalFn<S extends Record<string, unknown>> = (
  state: S,
) => EdgeTarget | Promise<EdgeTarget>;

type Node<S extends Record<string, unknown>> = {
  name: NodeName;
  fn: StepFn<S>;
};

type Edge =
  | { from: NodeName; to: EdgeTarget; kind: "static" }
  | { from: NodeName; kind: "conditional"; fn: ConditionalFn<any> };

export type GraphRunOptions = {
  maxSteps?: number; // safety cap, default 30
  signal?: AbortSignal;
};

export type GraphRunResult<S> = {
  state: S;
  path: NodeName[]; // ordered list of steps that executed
  workflowId: string | null;
};

export class Graph<S extends Record<string, unknown> = Record<string, unknown>> {
  private nodes = new Map<NodeName, Node<S>>();
  private edges: Edge[] = [];
  private entry: NodeName | null = null;

  // Register a step. The function receives the current state and returns
  // a partial state object that gets merged in.
  step(name: NodeName, fn: StepFn<S>): this {
    if (this.nodes.has(name)) {
      throw new Error(`graph: step "${name}" already defined`);
    }
    if (name === START || name === END) {
      throw new Error(`graph: "${name}" is a reserved name`);
    }
    this.nodes.set(name, { name, fn });
    return this;
  }

  // Convenience: wire an Agent in as a step. Reads `input` from a state
  // field (default "input"), runs the agent, writes the final text into
  // a target state field (default the step name).
  agent(
    name: NodeName,
    agent: Agent,
    options?: {
      inputFrom?: keyof S | string;
      outputTo?: keyof S | string;
      systemFrom?: keyof S | string;
    },
  ): this {
    const inputFrom = (options?.inputFrom ?? "input") as string;
    const outputTo = (options?.outputTo ?? name) as string;
    return this.step(name, async (state, ctx) => {
      const input = state[inputFrom as keyof S];
      if (typeof input !== "string" || !input) {
        throw new Error(
          `graph step "${name}": expected state["${inputFrom}"] to be a non-empty string`,
        );
      }
      const events = agent.run(input, {
        workflowId: ctx.workflowId,
        signal: ctx.signal,
      });
      const output = await collectFinalOutput(events);
      return { [outputTo]: output } as Partial<S>;
    });
  }

  // Static edge from `from` to `to`. Use END to mark a terminal state.
  edge(from: NodeName, to: EdgeTarget): this {
    this.assertNode(from);
    if (to !== END) this.assertNode(to);
    this.edges.push({ from, to, kind: "static" });
    return this;
  }

  // Conditional edge: when a step finishes, evaluate `fn(state)` to
  // decide where to go next. Multiple conditionals from the same source
  // are not allowed; static and conditional from the same source are
  // mutually exclusive.
  conditional(from: NodeName, fn: ConditionalFn<S>): this {
    this.assertNode(from);
    if (this.edges.some((e) => e.from === from)) {
      throw new Error(
        `graph: step "${from}" already has an outgoing edge; remove it before adding a conditional`,
      );
    }
    this.edges.push({ from, kind: "conditional", fn });
    return this;
  }

  // Mark the entry point. If unset, the first registered step is used.
  start(name: NodeName): this {
    this.assertNode(name);
    this.entry = name;
    return this;
  }

  // Execute the graph against an initial state.
  async run(initial: S, options?: GraphRunOptions): Promise<GraphRunResult<S>> {
    const maxSteps = Math.max(1, options?.maxSteps ?? 30);
    const entry = this.entry ?? this.nodes.keys().next().value;
    if (!entry) {
      throw new Error("graph: no steps defined");
    }

    let state: S = { ...initial };
    let workflowId: string | null = null;
    const path: NodeName[] = [];
    let current: NodeName | typeof END = entry;

    for (let i = 0; i < maxSteps; i++) {
      if (options?.signal?.aborted) {
        throw new Error("graph run aborted");
      }
      if (current === END) {
        return { state, path, workflowId };
      }

      const node = this.nodes.get(current)!;
      const ctx: StepContext = {
        workflowId: workflowId ?? "",
        stepName: node.name,
        signal: options?.signal,
      };
      path.push(node.name);

      const patch = await node.fn(state, ctx);
      if (patch && typeof patch === "object") {
        state = { ...state, ...patch };
      }

      // First step's workflowId discovery: if the step ran a Relay agent,
      // its workflowId is now on the most recent run. We don't have a
      // direct hook for that without intercepting the Agent's events, so
      // we leave workflowId null for pure-function steps. For
      // `agent()`-defined steps we set it via the AgentRunCollector below.
      const wf = patch && (patch as { __workflowId?: string }).__workflowId;
      if (wf && !workflowId) {
        workflowId = wf;
        // Strip the bookkeeping field from state.
        delete (state as Record<string, unknown>)["__workflowId"];
      }

      const next = await this.nextFrom(current, state);
      if (next === undefined) {
        throw new Error(
          `graph: step "${current}" finished but has no outgoing edge (use .edge(from, END) to terminate)`,
        );
      }
      current = next;
    }
    throw new Error(`graph: exceeded maxSteps (${maxSteps}); possible cycle`);
  }

  private async nextFrom(from: NodeName, state: S): Promise<NodeName | typeof END | undefined> {
    const matching = this.edges.find((e) => e.from === from);
    if (!matching) return undefined;
    if (matching.kind === "static") return matching.to;
    return matching.fn(state);
  }

  private assertNode(name: NodeName): void {
    if (!this.nodes.has(name)) {
      throw new Error(`graph: unknown step "${name}"`);
    }
  }

  // For debugging / future dashboard rendering.
  describe(): {
    entry: NodeName | null;
    steps: NodeName[];
    edges: Array<{ from: NodeName; kind: string; to?: EdgeTarget }>;
  } {
    return {
      entry: this.entry,
      steps: [...this.nodes.keys()],
      edges: this.edges.map((e) =>
        e.kind === "static"
          ? { from: e.from, kind: "static", to: e.to }
          : { from: e.from, kind: "conditional" },
      ),
    };
  }
}

// Helper that drains an Agent's event stream into the final text output.
// Exported for users who want to wire their own steps without using
// `.agent()`.
export async function collectFinalOutput(
  events: AsyncIterable<AgentEvent>,
): Promise<string> {
  const parts: string[] = [];
  let final: string | null = null;
  for await (const ev of events) {
    if (ev.type === "token") {
      parts.push(ev.text);
    } else if (ev.type === "done") {
      final = ev.output;
    } else if (ev.type === "error") {
      throw new Error(`agent step failed: ${ev.message}`);
    }
  }
  return final ?? parts.join("");
}

// Re-export createAgent so users can do everything graph-related from one
// import (`@relayhq/sdk` already exports it from agent.js).
export { createAgent };

File: sdks/python/relayhq/graph.py
"""Graph — declarative multi-step workflows.

Lightweight orchestrator built on Agent + subagent + the run-linking
infrastructure. Pure SDK, no server changes — the server still sees the
resulting runs as a tree linked by workflow_id.
"""

from __future__ import annotations

import inspect
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Dict, List, Optional, Union

from .agent import Agent
from .types import AgentEvent


START = "__start__"
END = "__end__"

# Step function signature: (state, ctx) -> partial state (sync or async).
StepFn = Callable[
    [Dict[str, Any], "StepContext"], Union[Dict[str, Any], Awaitable[Dict[str, Any]]]
]

# Conditional: receives state, returns next step name or END.
ConditionalFn = Callable[[Dict[str, Any]], Union[str, Awaitable[str]]]


@dataclass
class StepContext:
    """Per-step context. workflow_id is empty before the first agent
    step has discovered it from the server's x-workflow-id header."""

    workflow_id: str
    step_name: str


@dataclass
class _Node:
    name: str
    fn: StepFn


@dataclass
class _StaticEdge:
    source: str
    target: str  # may be END


@dataclass
class _ConditionalEdge:
    source: str
    fn: ConditionalFn


@dataclass
class GraphRunResult:
    state: Dict[str, Any]
    path: List[str] = field(default_factory=list)
    workflow_id: Optional[str] = None


class Graph:
    """Declarative graph of steps. Wire steps with .step()/.agent(),
    connect them with .edge()/.conditional(), then .run() with an initial
    state.

    Example:

        researcher = create_agent(model="gpt-4o", tools=[web_search])
        writer = create_agent(model="claude-sonnet-4-6")

        graph = (
            Graph()
              .agent("research", researcher,
                     input_from="topic", output_to="research")
              .agent("write", writer,
                     input_from="research", output_to="draft")
              .edge("research", "write")
              .edge("write", END)
              .start("research")
        )
        result = await graph.run({"topic": "AI agents"})
    """

    def __init__(self) -> None:
        self._nodes: Dict[str, _Node] = {}
        self._edges: List[Union[_StaticEdge, _ConditionalEdge]] = []
        self._entry: Optional[str] = None

    def step(self, name: str, fn: StepFn) -> "Graph":
        if name in self._nodes:
            raise ValueError(f'graph: step "{name}" already defined')
        if name in (START, END):
            raise ValueError(f'graph: "{name}" is a reserved name')
        self._nodes[name] = _Node(name=name, fn=fn)
        return self

    def agent(
        self,
        name: str,
        agent: Agent,
        *,
        input_from: str = "input",
        output_to: Optional[str] = None,
    ) -> "Graph":
        """Wire an Agent as a step. Reads input from state[input_from]
        and writes the final text to state[output_to] (defaults to name)."""
        out_key = output_to or name

        async def _runner(state: Dict[str, Any], ctx: StepContext) -> Dict[str, Any]:
            prompt = state.get(input_from)
            if not isinstance(prompt, str) or not prompt:
                raise RuntimeError(
                    f'graph step "{name}": expected state["{input_from}"] to be a non-empty string'
                )
            events = agent.run(
                prompt,
                workflow_id=ctx.workflow_id or None,
            )
            output = await collect_final_output(events)
            return {out_key: output}

        return self.step(name, _runner)

    def edge(self, source: str, target: str) -> "Graph":
        self._assert_node(source)
        if target != END:
            self._assert_node(target)
        self._edges.append(_StaticEdge(source=source, target=target))
        return self

    def conditional(self, source: str, fn: ConditionalFn) -> "Graph":
        self._assert_node(source)
        if any(e.source == source for e in self._edges):
            raise ValueError(
                f'graph: step "{source}" already has an outgoing edge; remove it before adding a conditional'
            )
        self._edges.append(_ConditionalEdge(source=source, fn=fn))
        return self

    def start(self, name: str) -> "Graph":
        self._assert_node(name)
        self._entry = name
        return self

    async def run(
        self,
        initial: Dict[str, Any],
        *,
        max_steps: int = 30,
    ) -> GraphRunResult:
        if not self._nodes:
            raise RuntimeError("graph: no steps defined")
        entry = self._entry or next(iter(self._nodes.keys()))

        state = dict(initial)
        path: List[str] = []
        workflow_id: Optional[str] = None
        current: str = entry

        for _ in range(max(1, max_steps)):
            if current == END:
                return GraphRunResult(state=state, path=path, workflow_id=workflow_id)
            node = self._nodes[current]
            path.append(node.name)
            ctx = StepContext(
                workflow_id=workflow_id or "",
                step_name=node.name,
            )
            result = node.fn(state, ctx)
            if inspect.isawaitable(result):
                result = await result
            if isinstance(result, dict):
                state.update(result)
            nxt = await self._next_from(current, state)
            if nxt is None:
                raise RuntimeError(
                    f'graph: step "{current}" finished but has no outgoing edge'
                )
            current = nxt

        raise RuntimeError(
            f"graph: exceeded max_steps ({max_steps}); possible cycle"
        )

    async def _next_from(self, source: str, state: Dict[str, Any]) -> Optional[str]:
        for edge in self._edges:
            if edge.source != source:
                continue
            if isinstance(edge, _StaticEdge):
                return edge.target
            result = edge.fn(state)
            if inspect.isawaitable(result):
                result = await result
            return result
        return None

    def _assert_node(self, name: str) -> None:
        if name not in self._nodes:
            raise ValueError(f'graph: unknown step "{name}"')

    def describe(self) -> Dict[str, Any]:
        return {
            "entry": self._entry,
            "steps": list(self._nodes.keys()),
            "edges": [
                {"from": e.source, "kind": "static", "to": e.target}
                if isinstance(e, _StaticEdge)
                else {"from": e.source, "kind": "conditional"}
                for e in self._edges
            ],
        }


async def collect_final_output(events: Any) -> str:
    """Drain an Agent event stream into the final text output."""
    parts: List[str] = []
    final: Optional[str] = None
    async for ev in events:
        if ev.get("type") == "token":
            parts.append(ev.get("text", ""))
        elif ev.get("type") == "done":
            final = ev.get("output")
        elif ev.get("type") == "error":
            raise RuntimeError(f"agent step failed: {ev.get('message')}")
    return final if final is not None else "".join(parts)
